File: src/store/roleState.ts
import { create } from "zustand";

import { api } from "../api/tauri";

export type RoleGender = "male" | "female";

/** Semen / fluid state under `nsfw.semen` (English keys). */
export interface RoleNsfwSemen {
  /** Male: semen texture / quality (text). */
  texture?: string;
  /** Female: external residue on body (text). */
  exterior?: string;
  /** Female: swallowed volume (ml). */
  swallowed?: number;
  /** Female: vaginal retention (ml). */
  vaginal?: number;
  /** Female: anal retention (ml). */
  anal?: number;
}

/** Structured NSFW block on a role card (English keys). */
export interface RoleNsfw {
  arousal?: number;
  wetness?: number;
  status?: string;
  sensitive_spots?: string[];
  semen?: RoleNsfwSemen;
  [key: string]: unknown;
}

/** One named meter rendered as a bar (e.g. 体力 80/100). */
export interface RoleMeter {
  value: number;
  max?: number;
}

/** A character on the state board. Free-form by design; only `id` is required.
 * The recommended numeric-first shape mirrors the backend `RoleState` tool. */
export interface Role {
  id: string;
  name?: string;
  /** `"male"` | `"female"` — drives semen field rendering. */
  gender?: RoleGender;
  location?: string;
  mood?: string;
  outfit?: string;
  /** Physical overview (≤100 Chinese chars); includes stature and genital scale. */
  appearance?: string;
  /** 0-100 scalar attributes → radar polygon. */
  attributes?: Record<string, number>;
  /** { value, max } gauges → bars. */
  meters?: Record<string, RoleMeter | number>;
  /** Short status chips. */
  tags?: string[];
  /** Explicit body / arousal state — see [`RoleNsfw`]. */
  nsfw?: RoleNsfw;
  [key: string]: unknown;
}

/** Shape of a `RoleState` tool result emitted over `gen://tool`. */
export interface RoleStateOp {
  op: "get" | "create" | "update" | "delete";
  id?: string;
  role?: Role;
  roles?: Role[];
  removed?: boolean;
}

/** Female semen volume keys (ml). */
export const SEMEN_ML_KEYS = ["swallowed", "vaginal", "anal"] as const;

const LEGACY_SEMEN_ML: Record<(typeof SEMEN_ML_KEYS)[number], string> = {
  swallowed: "吞精",
  vaginal: "阴道",
  anal: "肛门",
};

/** Read `nsfw.semen`, falling back to legacy `nsfw.精液`. */
export function resolveSemen(nsfw: RoleNsfw | undefined): RoleNsfwSemen | undefined {
  if (!nsfw) return undefined;
  const block = nsfw.semen ?? (nsfw as Record<string, unknown>)["精液"];
  if (block && typeof block === "object") return block as RoleNsfwSemen;
  return undefined;
}

/** Read `appearance`, with legacy top-level `外表` fallback. */
export function resolveAppearance(role: Role): string | null {
  const raw = role.appearance ?? (role as Record<string, unknown>)["外表"];
  if (typeof raw !== "string") return null;
  const trimmed = raw.trim();
  return trimmed || null;
}

/** Normalised gender from role (English keys only). */
export function resolveGender(role: Role): RoleGender | undefined {
  const g = role.gender;
  return g === "male" || g === "female" ? g : undefined;
}

/** Read a female semen ml field (new or legacy Chinese key). */
export function semenMl(
  semen: RoleNsfwSemen | undefined,
  key: (typeof SEMEN_ML_KEYS)[number],
): number | undefined {
  if (!semen) return undefined;
  const legacyKey = LEGACY_SEMEN_ML[key];
  const raw = semen[key] ?? (semen as Record<string, unknown>)[legacyKey];
  return typeof raw === "number" && Number.isFinite(raw) ? raw : undefined;
}

/** Read semen text: `texture` (male) or `exterior` (female), with legacy `外表`. */
export function semenText(
  semen: RoleNsfwSemen | undefined,
  kind: "texture" | "exterior",
): string | null {
  if (!semen) return null;
  const legacyExterior = (semen as Record<string, unknown>)["外表"];
  const raw =
    kind === "texture"
      ? semen.texture
      : semen.exterior ?? (typeof legacyExterior === "string" ? legacyExterior : undefined);
  if (typeof raw !== "string") return null;
  const trimmed = raw.trim();
  return trimmed || null;
}

/** NSFW scalar pairs [key, value] — English first, legacy Chinese fallback. */
export function nsfwScalars(nsfw: RoleNsfw): Array<[string, number]> {
  const pairs: Array<[string, number]> = [];
  const arousal =
    typeof nsfw.arousal === "number"
      ? nsfw.arousal
      : typeof (nsfw as Record<string, unknown>)["兴奋度"] === "number"
        ? ((nsfw as Record<string, unknown>)["兴奋度"] as number)
        : undefined;
  const wetness =
    typeof nsfw.wetness === "number"
      ? nsfw.wetness
      : typeof (nsfw as Record<string, unknown>)["湿润度"] === "number"
        ? ((nsfw as Record<string, unknown>)["湿润度"] as number)
        : undefined;
  if (typeof arousal === "number") pairs.push(["arousal", arousal]);
  if (typeof wetness === "number") pairs.push(["wetness", wetness]);
  return pairs;
}

export function nsfwStatus(nsfw: RoleNsfw): string | null {
  const raw = nsfw.status ?? (nsfw as Record<string, unknown>)["状态"];
  return typeof raw === "string" && raw.trim() ? raw.trim() : null;
}

export function nsfwSensitiveSpots(nsfw: RoleNsfw): string[] {
  const raw = nsfw.sensitive_spots ?? (nsfw as Record<string, unknown>)["敏感点"];
  return Array.isArray(raw) ? raw.map(String) : [];
}

/** Project sessions share `project_id`; standalone sessions use session id. */
export function resolveRoleStateScope(session: {
  id: string;
  project_id?: string | null;
}): string {
  return session.project_id ?? session.id;
}

interface RoleStateStore {
  orderByScope: Record<string, string[]>;
  rolesByScope: Record<string, Record<string, Role>>;

  applyOp: (scopeId: string, op: RoleStateOp) => void;
  setRoles: (scopeId: string, roles: Role[]) => void;
  loadLatest: (sessionId: string, scopeId: string) => Promise<void>;
  updateRole: (sessionId: string, scopeId: string, role: Role) => Promise<Role>;
  deleteRole: (sessionId: string, scopeId: string, id: string) => Promise<boolean>;
  rolesOf: (scopeId: string | null | undefined) => Role[];
}

function dedupeOrder(order: string[]): string[] {
  return Array.from(new Set(order));
}

export const useRoleState = create<RoleStateStore>((set, get) => ({
  orderByScope: {},
  rolesByScope: {},

  setRoles: (scopeId, roles) => {
    const map: Record<string, Role> = {};
    const order: string[] = [];
    for (const r of roles) {
      if (!r || typeof r.id !== "string") continue;
      map[r.id] = r;
      order.push(r.id);
    }
    set((s) => ({
      rolesByScope: { ...s.rolesByScope, [scopeId]: map },
      orderByScope: { ...s.orderByScope, [scopeId]: dedupeOrder(order) },
    }));
  },

  applyOp: (scopeId, op) => {
    if (op.op === "get") {
      if (Array.isArray(op.roles)) get().setRoles(scopeId, op.roles);
      return;
    }
    set((s) => {
      const map = { ...(s.rolesByScope[scopeId] ?? {}) };
      let order = [...(s.orderByScope[scopeId] ?? [])];

      if (op.op === "create" && op.role && typeof op.role.id === "string") {
        map[op.role.id] = op.role;
        if (!order.includes(op.role.id)) order.push(op.role.id);
      } else if (op.op === "update" && op.role && typeof op.role.id === "string") {
        map[op.role.id] = op.role;
        if (!order.includes(op.role.id)) order.push(op.role.id);
      } else if (op.op === "delete" && typeof op.id === "string") {
        delete map[op.id];
        order = order.filter((id) => id !== op.id);
      }

      return {
        rolesByScope: { ...s.rolesByScope, [scopeId]: map },
        orderByScope: { ...s.orderByScope, [scopeId]: order },
      };
    });
  },

  loadLatest: async (_sessionId, scopeId) => {
    try {
      const roles = await api.getRoleStates(_sessionId);
      get().setRoles(scopeId, Array.isArray(roles) ? roles : []);
    } catch (e) {
      console.warn("[roleState] loadLatest failed", e);
    }
  },

  updateRole: async (sessionId, scopeId, role) => {
    const updated = await api.updateRoleState(sessionId, role);
    get().applyOp(scopeId, { op: "update", id: updated.id, role: updated });
    return updated;
  },

  deleteRole: async (sessionId, scopeId, id) => {
    const result = await api.deleteRoleState(sessionId, id);
    if (result.removed) {
      get().applyOp(scopeId, { op: "delete", id });
    }
    return result.removed;
  },

  rolesOf: (scopeId) => {
    if (!scopeId) return [];
    const map = get().rolesByScope[scopeId];
    const order = get().orderByScope[scopeId];
    if (!map || !order) return [];
    return order.map((id) => map[id]).filter(Boolean) as Role[];
  },
}));

File: src-tauri/src/ai/agent/tools/role_state.rs
//! Incremental role-state tool.
//!
//! Drives the project-scoped "character state board" (standalone sessions fall
//! back to session scope). Unlike [`super::todo`] the store is **shared** across
//! every agent run that belongs to the same project (or standalone session),
//! can read what previous turns established and only emit the *delta*.
//!
//! Semantics are deliberately incremental (NEVER full-replace):
//!
//! | action   | required fields                | description                              |
//! |----------|--------------------------------|------------------------------------------|
//! | `get`    | —                              | Return every role currently on the board |
//! | `create` | `id`, `role`                   | Insert a brand-new role (stable `id`)    |
//! | `update` | `id`, (`set` and/or `unset`)   | Field-level dot-path edits / deletions   |
//! | `delete` | `id`                           | Remove one role entirely                 |
//!
//! `set` is a flat map of dot-paths → values, e.g.
//! `{ "attributes.好感": 80, "mood": "羞涩" }`. `unset` is an array of
//! dot-paths to remove, e.g. `["tags.1", "nsfw.敏感点"]`.
//!
//! Each role is a free-form JSON object; the recommended shape (encoded in
//! the schema description) favours numeric/structured fields so the UI can
//! render polygons/bars instead of walls of text:
//!
//! ```json
//! {
//!   "id": "rin", "name": "凛", "gender": "female",
//!   "location": "…", "mood": "…", "outfit": "…",
//!   "appearance": "高挑纤细，D罩杯丰满，阴唇粉嫩紧致",
//!   "attributes": { "好感": 72, "信任": 55 },
//!   "meters": { "体力": { "value": 80, "max": 100 } },
//!   "tags": ["害羞"],
//!   "nsfw": {
//!     "arousal": 40, "wetness": 55, "status": "迷离",
//!     "sensitive_spots": ["颈部"],
//!     "semen": {
//!       "exterior": "脸颊与胸前少量白浊",
//!       "swallowed": 12.5, "vaginal": 8.0, "anal": 0
//!     }
//!   }
//! }
//! ```
//! `gender` is `"male"` or `"female"` (required on `create`).
//! `appearance` — brief physical overview (≤100 Chinese characters). Set on
//! `create`; update only when the prose establishes or changes body traits.
//! Include stature / build and gender-specific genital scale, e.g. female
//! breast cup / size, male penis length & girth.
//! Semen fields under `nsfw.semen` depend on gender — use English keys only:
//! - **male** → `texture` (TEXT: semen quality / viscosity / warmth).
//! - **female** → `exterior` (TEXT: external residue) plus `swallowed` /
//!   `vaginal` / `anal` as millilitre amounts (ml, NOT 0-100).

use std::collections::HashMap;
use std::sync::{Arc, Mutex};

use serde_json::{json, Map, Value};

use crate::ai::agent::tools::{Tool, ToolFuture, ToolInvocation, ToolResult, ToolSpec};
use crate::error::{AppError, AppResult};

pub const TOOL_NAME: &str = "RoleState";

/// Scope-scoped, shared store of role boards. Lives on `AppState` so the
/// `role-state` sub-agent, the persistence layer and the Tauri commands all
/// see the same in-memory truth.
#[derive(Debug, Default)]
pub struct RoleStateStore {
    /// scope_id (project_id or session_id) → ordered role list.
    boards: Mutex<HashMap<String, Vec<Value>>>,
}

impl RoleStateStore {
    pub fn new() -> Self {
        Self::default()
    }

    /// Snapshot every role for a scope as a JSON array (insertion order).
    pub fn snapshot(&self, scope_id: &str) -> Vec<Value> {
        self.boards
            .lock()
            .ok()
            .and_then(|g| g.get(scope_id).cloned())
            .unwrap_or_default()
    }

    /// Replace a scope's roles wholesale. Used when loading a persisted
    /// snapshot back into memory (session open / rollback).
    pub fn load(&self, scope_id: &str, roles: Vec<Value>) {
        if let Ok(mut g) = self.boards.lock() {
            if roles.is_empty() {
                g.remove(scope_id);
            } else {
                g.insert(scope_id.to_string(), roles);
            }
        }
    }

    /// Drop a scope's roles entirely (e.g. when no snapshot remains).
    pub fn clear(&self, scope_id: &str) {
        if let Ok(mut g) = self.boards.lock() {
            g.remove(scope_id);
        }
    }

    fn create(&self, scope_id: &str, id: &str, role: Value) -> AppResult<Value> {
        let mut role = match role {
            Value::Object(m) => m,
            _ => {
                return Err(AppError::Invalid(
                    "RoleState create: `role` must be an object".into(),
                ))
            }
        };
        role.insert("id".into(), Value::String(id.to_string()));
        let role = Value::Object(role);

        let mut g = self
            .boards
            .lock()
            .map_err(|_| AppError::Other("RoleState: store lock poisoned".into()))?;
        let list = g.entry(scope_id.to_string()).or_default();
        match list.iter_mut().find(|r| role_id(r) == Some(id)) {
            Some(slot) => {
                // Idempotent create: overwrite the existing entry.
                *slot = role.clone();
            }
            None => list.push(role.clone()),
        }
        Ok(role)
    }

    fn update(
        &self,
        scope_id: &str,
        id: &str,
        set: Option<&Map<String, Value>>,
        unset: &[String],
    ) -> AppResult<Value> {
        let mut g = self
            .boards
            .lock()
            .map_err(|_| AppError::Other("RoleState: store lock poisoned".into()))?;
        let list = g
            .get_mut(scope_id)
            .ok_or_else(|| AppError::Invalid("RoleState update: no roles for scope".into()))?;
        let role = list
            .iter_mut()
            .find(|r| role_id(r) == Some(id))
            .ok_or_else(|| AppError::Invalid(format!("RoleState update: unknown role id {id:?}")))?;

        if let Some(set) = set {
            for (path, value) in set {
                set_dot_path(role, path, value.clone());
            }
        }
        for path in unset {
            unset_dot_path(role, path);
        }
        Ok(role.clone())
    }

    /// Full-object replace for a role (UI edit). Preserves list order; errors if
    /// the id is missing from the board.
    pub fn replace(&self, scope_id: &str, id: &str, role: Value) -> AppResult<Value> {
        let mut role = match role {
            Value::Object(m) => m,
            _ => {
                return Err(AppError::Invalid(
                    "RoleState replace: `role` must be an object".into(),
                ))
            }
        };
        role.insert("id".into(), Value::String(id.to_string()));
        let role = Value::Object(role);

        let mut g = self
            .boards
            .lock()
            .map_err(|_| AppError::Other("RoleState: store lock poisoned".into()))?;
        let list = g
            .get_mut(scope_id)
            .ok_or_else(|| AppError::Invalid("RoleState replace: no roles for scope".into()))?;
        let slot = list
            .iter_mut()
            .find(|r| role_id(r) == Some(id))
            .ok_or_else(|| {
                AppError::Invalid(format!("RoleState replace: unknown role id {id:?}"))
            })?;
        *slot = role.clone();
        Ok(role)
    }

    pub fn delete(&self, scope_id: &str, id: &str) -> AppResult<bool> {
        let mut g = self
            .boards
            .lock()
            .map_err(|_| AppError::Other("RoleState: store lock poisoned".into()))?;
        let Some(list) = g.get_mut(scope_id) else {
            return Ok(false);
        };
        let before = list.len();
        list.retain(|r| role_id(r) != Some(id));
        let removed = list.len() < before;
        if list.is_empty() {
            g.remove(scope_id);
        }
        Ok(removed)
    }
}

fn role_id(role: &Value) -> Option<&str> {
    role.get("id").and_then(Value::as_str)
}

/// Set a value at a dot-path, creating intermediate objects as needed.
/// Pure-numeric path segments index into arrays when the parent is one.
fn set_dot_path(root: &mut Value, path: &str, value: Value) {
    let parts: Vec<&str> = path.split('.').collect();
    let mut cur = root;
    for (i, part) in parts.iter().enumerate() {
        let last = i == parts.len() - 1;
        if last {
            assign_segment(cur, part, value);
            return;
        }
        cur = descend_segment(cur, part);
    }
}

fn assign_segment(parent: &mut Value, seg: &str, value: Value) {
    match parent {
        Value::Array(arr) => {
            if let Ok(idx) = seg.parse::<usize>() {
                if idx < arr.len() {
                    arr[idx] = value;
                } else {
                    arr.push(value);
                }
            }
        }
        Value::Object(map) => {
            map.insert(seg.to_string(), value);
        }
        other => {
            let mut map = Map::new();
            map.insert(seg.to_string(), value);
            *other = Value::Object(map);
        }
    }
}

fn descend_segment<'a>(parent: &'a mut Value, seg: &str) -> &'a mut Value {
    if !parent.is_object() && !parent.is_array() {
        *parent = Value::Object(Map::new());
    }
    match parent {
        Value::Object(map) => map
            .entry(seg.to_string())
            .or_insert_with(|| Value::Object(Map::new())),
        Value::Array(arr) => {
            let idx = seg.parse::<usize>().unwrap_or(0);
            while arr.len() <= idx {
                arr.push(Value::Object(Map::new()));
            }
            &mut arr[idx]
        }
        // Unreachable after the coercion above, but the borrow checker
        // wants every arm to yield a reference.
        other => other,
    }
}

fn unset_dot_path(root: &mut Value, path: &str) {
    let parts: Vec<&str> = path.split('.').collect();
    let mut cur = root;
    for (i, part) in parts.iter().enumerate() {
        let last = i == parts.len() - 1;
        if last {
            match cur {
                Value::Object(map) => {
                    map.remove(*part);
                }
                Value::Array(arr) => {
                    if let Ok(idx) = part.parse::<usize>() {
                        if idx < arr.len() {
                            arr.remove(idx);
                        }
                    }
                }
                _ => {}
            }
            return;
        }
        cur = match cur {
            Value::Object(map) => match map.get_mut(*part) {
                Some(v) => v,
                None => return,
            },
            Value::Array(arr) => match part.parse::<usize>().ok().and_then(|i| arr.get_mut(i)) {
                Some(v) => v,
                None => return,
            },
            _ => return,
        };
    }
}

/// The `RoleState` tool. Holds a shared [`RoleStateStore`] reference and the
/// session id is read off [`ToolUseContext`] at execution time.
#[derive(Clone)]
pub struct RoleStateTool {
    spec: ToolSpec,
    store: Arc<RoleStateStore>,
}

impl RoleStateTool {
    pub fn new(store: Arc<RoleStateStore>) -> Self {
        Self {
            store,
            spec: ToolSpec {
                name: TOOL_NAME.to_string(),
                description: "\
Maintain the per-conversation character state board. STRICTLY INCREMENTAL — \
never re-send a whole role you have already created.\n\n\
━━━ WORKFLOW ━━━\n\
1. Call `get` FIRST to see which roles already exist and their current fields.\n\
2. For a character that does NOT exist yet → `create` with a stable, \
lowercase ascii `id` (e.g. \"rin\", \"alice\") plus the initial `role` object.\n\
3. For a character that already exists → `update` ONLY the fields that changed, \
using dot-paths in `set` / `unset`. Do NOT recreate it.\n\
4. For a character that left the scene for good → `delete`.\n\n\
━━━ ROLE OBJECT SHAPE (prefer numbers over prose; English keys for nsfw) ━━━\n\
{\n  \"name\": \"凛\", \"gender\": \"female\",\n  \"location\": \"…\", \"mood\": \"…\", \"outfit\": \"…\",\n  \"appearance\": \"高挑纤细，D罩杯丰满，阴唇粉嫩紧致\",\n  \"attributes\": { \"好感\": 72, \"信任\": 55 },   // 0-100, radar polygon\n  \"meters\": { \"体力\": { \"value\": 80, \"max\": 100 } }, // bars\n  \"tags\": [\"害羞\"],\n  \"nsfw\": {\n    \"arousal\": 40, \"wetness\": 55, \"status\": \"迷离\",\n    \"sensitive_spots\": [\"颈部\"],\n    \"semen\": { \"exterior\": \"小腹至大腿漫溢\", \"swallowed\": 850, \"vaginal\": 2400, \"anal\": 0 }\n  }\n}\n\
`gender` — REQUIRED on `create`: `\"male\"` or `\"female\"`.\n\
`appearance` — physical overview, ≤100 Chinese characters. REQUIRED on `create` \
when the prose describes the body; update only when traits change. Summarise \
stature / build AND gender-specific genital scale: female → breast size/cup; \
male → penis length & girth. Example male: \"精壮高挑，阴茎约18cm粗长\".\n\
`nsfw.semen` — gender-specific (update only what changed):\n\
  • **male** → `texture` (TEXT: semen quality — viscosity, warmth, colour, volume feel).\n\
  • **female** → `exterior` (TEXT: external residue on body) + `swallowed` / `vaginal` / \
`anal` as ml (can be large: horse/beast → 2000–8000 ml).\n\
Match FICTION scale, NOT clinical human averages. Use dot-paths, e.g. \
`nsfw.semen.vaginal`: 3200 or `nsfw.semen.texture`: \"浓稠、量大、温热\".\n\
Keep text fields brief. ALWAYS maintain `nsfw` (including `semen`) from the prose.\n\n\
━━━ INCREMENTAL EDIT EXAMPLES ━━━\n\
set: { \"nsfw.arousal\": 60, \"nsfw.semen.vaginal\": 1800, \"nsfw.semen.texture\": \"浓稠拉丝\" }\n\
unset: [\"tags.0\", \"nsfw.semen.anal\"]\n\n\
Do NOT narrate. Your visible text reply must be at most one short sentence."
                    .to_string(),
                schema: json!({
                    "type": "object",
                    "properties": {
                        "action": {
                            "type": "string",
                            "enum": ["get", "create", "update", "delete"],
                            "description": "The operation to perform."
                        },
                        "id": {
                            "type": "string",
                            "description": "Stable role id (required for create/update/delete)."
                        },
                        "role": {
                            "type": "object",
                            "description": "Full role object for `create`."
                        },
                        "set": {
                            "type": "object",
                            "description": "For `update`: map of dot-path → new value, e.g. {\"attributes.好感\": 80}."
                        },
                        "unset": {
                            "type": "array",
                            "items": { "type": "string" },
                            "description": "For `update`: dot-paths to remove, e.g. [\"tags.0\"]."
                        }
                    },
                    "required": ["action"]
                }),
                read_only: false,
                concurrency_safe: false,
            },
        }
    }
}

impl Tool for RoleStateTool {
    fn spec(&self) -> &ToolSpec {
        &self.spec
    }

    fn validate(&self, input: &Value) -> AppResult<()> {
        let action = input
            .get("action")
            .and_then(Value::as_str)
            .ok_or_else(|| AppError::Invalid("RoleState: `action` must be a string".into()))?;
        match action {
            "get" => {}
            "create" => {
                if input.get("id").and_then(Value::as_str).is_none() {
                    return Err(AppError::Invalid(
                        "RoleState create: `id` must be a string".into(),
                    ));
                }
                if !input.get("role").map(Value::is_object).unwrap_or(false) {
                    return Err(AppError::Invalid(
                        "RoleState create: `role` must be an object".into(),
                    ));
                }
            }
            "update" => {
                if input.get("id").and_then(Value::as_str).is_none() {
                    return Err(AppError::Invalid(
                        "RoleState update: `id` must be a string".into(),
                    ));
                }
                let has_set = input.get("set").map(Value::is_object).unwrap_or(false);
                let has_unset = input.get("unset").map(Value::is_array).unwrap_or(false);
                if !has_set && !has_unset {
                    return Err(AppError::Invalid(
                        "RoleState update: provide `set` (object) and/or `unset` (array)".into(),
                    ));
                }
            }
            "delete" => {
                if input.get("id").and_then(Value::as_str).is_none() {
                    return Err(AppError::Invalid(
                        "RoleState delete: `id` must be a string".into(),
                    ));
                }
            }
            other => {
                return Err(AppError::Invalid(format!(
                    "RoleState: unknown action {other:?}; must be get|create|update|delete"
                )));
            }
        }
        Ok(())
    }

    fn execute<'a>(&'a self, invocation: ToolInvocation<'a>) -> ToolFuture<'a> {
        let store = self.store.clone();
        Box::pin(async move {
            let scope_id = invocation
                .context
                .role_state_scope_id
                .clone()
                .or_else(|| invocation.context.session_id.clone())
                .ok_or_else(|| {
                AppError::Invalid(
                    "RoleState: no session id on this run; cannot persist role state".into(),
                )
            })?;
            let input = invocation.input;
            let action = input
                .get("action")
                .and_then(Value::as_str)
                .unwrap_or("")
                .to_string();

            match action.as_str() {
                "get" => {
                    let roles = store.snapshot(&scope_id);
                    Ok(ToolResult::ok(json!({
                        "op": "get",
                        "roles": roles,
                    })))
                }
                "create" => {
                    let id = input.get("id").and_then(Value::as_str).unwrap_or("").to_string();
                    let role = input.get("role").cloned().unwrap_or(Value::Null);
                    let role = store.create(&scope_id, &id, role)?;
                    Ok(ToolResult::ok(json!({
                        "op": "create",
                        "id": id,
                        "role": role,
                    })))
                }
                "update" => {
                    let id = input.get("id").and_then(Value::as_str).unwrap_or("").to_string();
                    let set = input.get("set").and_then(Value::as_object);
                    let unset: Vec<String> = input
                        .get("unset")
                        .and_then(Value::as_array)
                        .map(|a| {
                            a.iter()
                                .filter_map(|v| v.as_str().map(str::to_string))
                                .collect()
                        })
                        .unwrap_or_default();
                    let role = store.update(&scope_id, &id, set, &unset)?;
                    Ok(ToolResult::ok(json!({
                        "op": "update",
                        "id": id,
                        "role": role,
                    })))
                }
                "delete" => {
                    let id = input.get("id").and_then(Value::as_str).unwrap_or("").to_string();
                    let removed = store.delete(&scope_id, &id)?;
                    Ok(ToolResult::ok(json!({
                        "op": "delete",
                        "id": id,
                        "removed": removed,
                    })))
                }
                other => Err(AppError::Invalid(format!(
                    "RoleState: unknown action {other:?}"
                ))),
            }
        })
    }
}
